Exception in poll causes crash of entire instance

If there's an unhandled exception in a poll, it causes the entire instance to crash.

Obviously it would be ideal if people writing poll plugins handled all exceptions, but imo we should deal with cases where people don't. Many times these exceptions are temporary (network issues, etc.) and won't happen again if the poll runs again. Right now, the whole process crashes and if we're lucky, sr3 sanity will restart it.

I think this section should be wrapped in a try/except:

https://github.com/MetPX/sarracenia/blob/6ad476d5d1438bfa101f5754ed11fbc0cccd5b00/sarracenia/flow/__init__.py#L1039-L1043

We have a try/except for [other callbacks](https://github.com/MetPX/sarracenia/blob/6ad476d5d1438bfa101f5754ed11fbc0cccd5b00/sarracenia/flow/__init__.py#L275-L329), so a poll should be the same. 

 
## Current Behaviour:

The entire instance crashes.

```
$ sr3 foreground poll/nasa_omisips_omto3
exchange: xs_NASA-TEST_ split: 5
.2024-07-21 20:15:18,288 [WARNING] 3055003 sarracenia.flow.poll __init__ post_exchange: xs_NASA-TEST is different from exchange: xs_bunnymaster. The settings need for multiple instances to share a poll.
2024-07-21 20:15:18,288 [INFO] 3055003 sarracenia.flow.poll __init__ adding poll plugin, because missing from: ['sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'sarracenia.flowcb.nodupe.disk.Disk', 'poll.nasa_omisips', 'log', 'sarracenia.flowcb.scheduled.poll.Poll']
2024-07-21 20:15:18,288 [WARNING] 3055003 sarracenia.flow.poll __init__ nodupe_ttl < fileAgeMax means some files could age out of the cache and be re-ingested ( see : https://github.com/MetPX/sarracenia/issues/904
2024-07-21 20:15:18,289 [INFO] 3055003 sarracenia.flow loadCallbacks flowCallback plugins to load: ['sarracenia.flowcb.post.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'sarracenia.flowcb.nodupe.disk.Disk', 'poll.nasa_omisips', 'log', 'sarracenia.flowcb.scheduled.poll.Poll', 'sarracenia.flowcb.poll.Poll']
2024-07-21 20:15:18,297 [INFO] 3055003 sarracenia.flowcb.nodupe.disk __init__ time_to_live=1209600,
2024-07-21 20:15:18,306 [INFO] 3055003 sarracenia.flowcb.log __init__ poll initialized with: logEvents: {'after_post', 'after_accept', 'on_housekeeping', 'after_work'},  logMessageDump: False
2024-07-21 20:15:18,309 [INFO] 3055003 sarracenia.flowcb.scheduled update_appointments for 2024-07-21 20:15:18.309413+00:00: []
2024-07-21 20:15:18,310 [INFO] 3055003 sarracenia.flow run callbacks loaded: ['sarracenia.flowcb.post.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'sarracenia.flowcb.nodupe.disk.Disk', 'poll.nasa_omisips', 'log', 'sarracenia.flowcb.scheduled.poll.Poll', 'sarracenia.flowcb.poll.Poll']
2024-07-21 20:15:18,310 [INFO] 3055003 sarracenia.flow run pid: 3055003 poll/nasa_omisips_omto3 instance: 0
2024-07-21 20:15:18,330 [INFO] 3055003 sarracenia.moth.amqp putSetup exchange declared: xs_NASA-TEST (as: amqp://bunnymaster@localhost/)
2024-07-21 20:15:18,330 [INFO] 3055003 sarracenia.flowcb.post.message on_start starting
2024-07-21 20:15:18,332 [INFO] 3055003 sarracenia.flowcb.scheduled.poll gather waiting for next poll
Traceback (most recent call last):
  File "/net/local/home/sunderlandr/sr3/sarracenia/instance.py", line 248, in <module>
    i.start()
  File "/net/local/home/sunderlandr/sr3/sarracenia/instance.py", line 239, in start
    self.running_instance.run()
  File "/net/local/home/sunderlandr/sr3/sarracenia/flow/__init__.py", line 486, in run
    self.gather()
  File "/net/local/home/sunderlandr/sr3/sarracenia/flow/__init__.py", line 1041, in gather
    new_incoming = plugin()
  File "/net/local/home/sunderlandr/.config/sr3/plugins/poll/nasa_omisips.py", line 61, in poll
    logger.info(hello_this_should_crash)
NameError: name 'hello_this_should_crash' is not defined
2024-07-21 20:15:18,399 3054998 [CRITICAL] root run_command subprocess.run failed err=Command '['/usr/bin/python3', '/net/local/home/sunderlandr/sr3/sarracenia/instance.py', '--no', '0', 'foreground', 'poll/nasa_omisips_omto3']' returned non-zero exit status 1.
```

## Behaviour with a try/except added:

```
sunderlandr@edcm-dirt22-4:~/sr3$ sr3 foreground poll/nasa_omisips_omto3
exchange: xs_NASA-TEST_ split: 5
.2024-07-21 21:24:28,308 [WARNING] 3063160 sarracenia.flow.poll __init__ post_exchange: xs_NASA-TEST is different from exchange: xs_bunnymaster. The settings need for multiple instances to share a poll.
2024-07-21 21:24:28,308 [INFO] 3063160 sarracenia.flow.poll __init__ adding poll plugin, because missing from: ['sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'sarracenia.flowcb.nodupe.disk.Disk', 'poll.nasa_omisips', 'log', 'sarracenia.flowcb.scheduled.poll.Poll']
2024-07-21 21:24:28,308 [WARNING] 3063160 sarracenia.flow.poll __init__ nodupe_ttl < fileAgeMax means some files could age out of the cache and be re-ingested ( see : https://github.com/MetPX/sarracenia/issues/904
2024-07-21 21:24:28,308 [INFO] 3063160 sarracenia.flow loadCallbacks flowCallback plugins to load: ['sarracenia.flowcb.post.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'sarracenia.flowcb.nodupe.disk.Disk', 'poll.nasa_omisips', 'log', 'sarracenia.flowcb.scheduled.poll.Poll', 'sarracenia.flowcb.poll.Poll']
2024-07-21 21:24:28,316 [INFO] 3063160 sarracenia.flowcb.nodupe.disk __init__ time_to_live=1209600,
2024-07-21 21:24:28,323 [INFO] 3063160 sarracenia.flowcb.log __init__ poll initialized with: logEvents: {'after_accept', 'after_work', 'on_housekeeping', 'after_post'},  logMessageDump: False
2024-07-21 21:24:28,325 [INFO] 3063160 sarracenia.flowcb.scheduled update_appointments for 2024-07-21 21:24:28.325443+00:00: []
2024-07-21 21:24:28,326 [INFO] 3063160 sarracenia.flow run callbacks loaded: ['sarracenia.flowcb.post.message.Message', 'sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'sarracenia.flowcb.nodupe.disk.Disk', 'poll.nasa_omisips', 'log', 'sarracenia.flowcb.scheduled.poll.Poll', 'sarracenia.flowcb.poll.Poll']
2024-07-21 21:24:28,326 [INFO] 3063160 sarracenia.flow run pid: 3063160 poll/nasa_omisips_omto3 instance: 0
2024-07-21 21:24:28,345 [INFO] 3063160 sarracenia.moth.amqp putSetup exchange declared: xs_NASA-TEST (as: amqp://bunnymaster@localhost/)
2024-07-21 21:24:28,345 [INFO] 3063160 sarracenia.flowcb.post.message on_start starting
2024-07-21 21:24:28,346 [INFO] 3063160 sarracenia.flowcb.scheduled.poll gather waiting for next poll
2024-07-21 21:24:28,346 [ERROR] 3063160 sarracenia.flow gather flowCallback plugin poll.nasa_omisips.Nasa_omisips.poll crashed: name 'hello_this_should_crash' is not defined
2024-07-21 21:24:28,846 [INFO] 3063160 sarracenia.flowcb.scheduled.poll gather waiting for next poll
```

or and with debug logging, it gives the details:

```
2024-07-21 21:25:33,048 [ERROR] 3063659 sarracenia.flow gather flowCallback plugin poll.nasa_omisips.Nasa_omisips.poll crashed: name 'hello_this_should_crash' is not defined
2024-07-21 21:25:33,049 [DEBUG] 3063659 sarracenia.flow gather details:
Traceback (most recent call last):
  File "/net/local/home/sunderlandr/sr3/sarracenia/flow/__init__.py", line 1042, in gather
    new_incoming = plugin()
  File "/net/local/home/sunderlandr/.config/sr3/plugins/poll/nasa_omisips.py", line 61, in poll
    logger.info(hello_this_should_crash)
NameError: name 'hello_this_should_crash' is not defined
```

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -328,6 +328,25 @@ class Flow:
                     logger.error( f'flowCallback plugin {p}/metricsReport crashed: {ex}' )
                     logger.debug( "details:", exc_info=True )
 
+    def _runCallbackPoll(self):
+        for plugin in self.plugins['poll']:
+            if self.o.logLevel.lower() == 'debug' :
+                new_incoming = plugin()
+                if len(new_incoming) > 0:
+                    self.worklist.incoming.extend(new_incoming)
+            else:
+                try:
+                    new_incoming = plugin()
+                    if len(new_incoming) > 0:
+                        self.worklist.incoming.extend(new_incoming)
+                except Exception as ex:
+                    try:
+                        logger.error(f'flowCallback plugin {plugin.__module__}.{plugin.__qualname__} crashed: {ex}' )
+                    except:
+                        # just in case
+                        logger.error(f'flowCallback plugin {plugin} crashed: {ex}' )
+                    logger.debug("details:", exc_info=True )
+
     def _runHousekeeping(self, now) -> float:
         """ Run housekeeping callbacks
             Return the time when housekeeping should be run next
@@ -1037,20 +1056,7 @@ class Flow:
             self.worklist.poll_catching_up = False
 
         if self.have_vip:
-            for plugin in self.plugins['poll']:
-                try:
-                    new_incoming = plugin()
-                    if len(new_incoming) > 0:
-                        self.worklist.incoming.extend(new_incoming)
-                except Exception as ex:
-                    try:
-                        logger.error(f'flowCallback plugin {plugin.__module__}.{plugin.__qualname__} crashed: {ex}' )
-                    except:
-                        # just in case
-                        logger.error(f'flowCallback plugin {plugin} crashed: {ex}' )
-                    logger.debug("details:", exc_info=True )
-
-
+            self._runCallbackPoll()
 
     def do(self) -> None:
 

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -1038,9 +1038,17 @@ class Flow:
 
         if self.have_vip:
             for plugin in self.plugins['poll']:
-                new_incoming = plugin()
-                if len(new_incoming) > 0:
-                    self.worklist.incoming.extend(new_incoming)
+                try:
+                    new_incoming = plugin()
+                    if len(new_incoming) > 0:
+                        self.worklist.incoming.extend(new_incoming)
+                except Exception as ex:
+                    try:
+                        logger.error(f'flowCallback plugin {plugin.__module__}.{plugin.__qualname__} crashed: {ex}' )
+                    except:
+                        # just in case
+                        logger.error(f'flowCallback plugin {plugin} crashed: {ex}' )
+                    logger.debug("details:", exc_info=True )
 
 
 

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -274,6 +274,16 @@ class Flow:
 
     def _runCallbacksWorklist(self, entry_point):
 
+        if hasattr(self, entry_point):
+            if self.o.logLevel.lower() == 'debug' :
+                eval( f"self.{entry_point}(self.worklist)")
+            else:
+                try:
+                    eval( f"self.{entry_point}(self.worklist)")
+                except Exception as ex:
+                    logger.error( f'flow {entry_point} crashed: {ex}' )
+                    logger.debug( "details:", exc_info=True )
+
         if hasattr(self, 'plugins') and (entry_point in self.plugins):
             for p in self.plugins[entry_point]:
                 if self.o.logLevel.lower() == 'debug' :
@@ -286,6 +296,18 @@ class Flow:
                         logger.debug( "details:", exc_info=True )
 
     def runCallbacksTime(self, entry_point):
+
+        if hasattr(self, entry_point):
+            if self.o.logLevel.lower() == 'debug' :
+                eval( f"self.{entry_point}()")
+            else:
+                try:
+                    logger.info( f'normal run of self.{entry_point}' )
+                    eval( f"self.{entry_point}()")
+                except Exception as ex:
+                    logger.error( f'flow {entry_point} crashed: {ex}' )
+                    logger.debug( "details:", exc_info=True )
+
         for p in self.plugins[entry_point]:
             if self.o.logLevel.lower() == 'debug' :
                 p()
@@ -302,6 +324,16 @@ class Flow:
         Expects the plugin to return a dictionary containing metrics, which is saved to ``self.metrics[plugin_name]``.
         """
         
+        if hasattr(self, "metricsReport"):
+            if self.o.logLevel.lower() == 'debug' :
+                self.metricsReport()
+            else:
+                try:
+                    self.metricsReport()
+                except Exception as ex:
+                    logger.error( f'flow metricsReport() crashed: {ex}' )
+                    logger.debug( "details:", exc_info=True )
+
         if 'transferConnected' in self.metrics['flow'] and self.metrics['flow']['transferConnected']: 
             now=nowflt()
             self.metrics['flow']['transferConnectTime'] += now - self.metrics['flow']['transferConnectStart']
@@ -329,6 +361,16 @@ class Flow:
                     logger.debug( "details:", exc_info=True )
 
     def _runCallbackPoll(self):
+        if hasattr(self, "Poll"):
+            if self.o.logLevel.lower() == 'debug' :
+                self.Poll()
+            else:
+                try:
+                    self.Poll()
+                except Exception as ex:
+                    logger.error( f'flow Poll crashed: {ex}' )
+                    logger.debug( "details:", exc_info=True )
+
         for plugin in self.plugins['poll']:
             if self.o.logLevel.lower() == 'debug' :
                 new_incoming = plugin()
@@ -352,6 +394,16 @@ class Flow:
             Return the time when housekeeping should be run next
         """
         logger.info(f'on_housekeeping pid: {os.getpid()} {self.o.component}/{self.o.config} instance: {self.o.no}')
+        if hasattr(self, "on_housekeeping"):
+            if self.o.logLevel.lower() == 'debug' :
+                self.on_housekeeping()
+            else:
+                try:
+                    self.on_housekeeping()
+                except Exception as ex:
+                    logger.error( f'flow on_housekeeping crashed: {ex}' )
+                    logger.debug( "details:", exc_info=True )
+
         self.runCallbacksTime('on_housekeeping')
         self.metricsFlowReset()
         self.metrics['flow']['last_housekeeping'] = now

--- a/sarracenia/flow/poll.py
+++ b/sarracenia/flow/poll.py
@@ -65,10 +65,6 @@ class Poll(Flow):
         if not 'scheduled' in ','.join(self.plugins['load']):
             self.plugins['load'].append('sarracenia.flowcb.scheduled.poll.Poll')
 
-        if not 'flowcb.poll.Poll' in ','.join(self.plugins['load']):
-            logger.info( f"adding poll plugin, because missing from: {self.plugins['load']}" ) 
-            self.plugins['load'].append('sarracenia.flowcb.poll.Poll')
-
         if options.vip:
             self.plugins['load'].insert( 0, 'sarracenia.flowcb.gather.message.Message')
 
@@ -80,3 +76,14 @@ class Poll(Flow):
         if not features['ftppoll']['present']:
             if hasattr( self.o, 'pollUrl' ) and ( self.o.pollUrl.startswith('ftp') ):
                 logger.critical( f"attempting to configure an FTP poll pollUrl={self.o.pollUrl}, but missing python modules: {' '.join(features['ftppoll']['modules_needed'])}" )
+
+    def on_start(self):
+
+        if 'poll' not in self.plugins or not self.plugins['poll']:
+            logger.info( f"adding built-in poll plugin, because no other poll provided from: {self.plugins['load']}" ) 
+
+            self.plugins['load'].append('sarracenia.flowcb.poll.Poll')
+            plugin = sarracenia.flowcb.load_library("sarracenia.flowcb.poll.Poll", self.o)
+            self.plugins['poll'] = [ getattr(plugin, 'poll') ]
+        else:
+            logger.info( f"not adding built-in poll, because already present: {self.plugins['poll']} " )

